### PR TITLE
PCHR-1746: Prevent multiple saves.

### PR DESCRIPTION
**Pull request by sleekvick**
# Multiple saves on multi-click

Because of network latency users can save regions and locations more than once.

# Preventing multiple saves with a boolean.

The solution is introducing a boolean variable to prevent the user from clicking the save button more than once.  

Also the label of the regions is used instead of the name to display the accurate names.

pchr-1746

### CODE REVIEW

File: contactaccessrights/js/src/access-rights/controllers/access-rights-modal-ctrl.js [hunk 1]
```diff
@@ -39,11 +41,14 @@ define([
        * Saves data and closes the modal
        */
       vm.submit = function () {
+        vm.hasSubmit = true;
         $q.all([persistValues('regions'), persistValues('locations')])
```
Comment (AkA84) [inline contactaccessrights/js/src/access-rights/controllers/access-rights-modal-ctrl.js]: I wonder if you could do the same with simply `.finally()`?
```js
$q.all()
  .then(function () {

  })
  .catch(function () {

  })
  .finally(function () {
    vm.submitting = false;
  });
```
not sure thought, you should test it

File: contactaccessrights/js/src/access-rights/controllers/access-rights-modal-ctrl.js [hunk 2]
```diff
@@ -9,6 +9,8 @@ define([
     function ($q, $modalInstance, Region, Location, Right) {
       var vm = this;
 
+      vm.hasSubmit = false;
```
Comment (AkA84) [inline contactaccessrights/js/src/access-rights/controllers/access-rights-modal-ctrl.js]: Please use a better name for the property. Something like `vm.submitting`;

File: contactaccessrights/js/src/access-rights/controllers/access-rights-modal-ctrl.js [hunk 3]
```diff
@@ -39,11 +41,14 @@ define([
        * Saves data and closes the modal
        */
       vm.submit = function () {
+        vm.hasSubmit = true;
```
Comment (AkA84) [inline contactaccessrights/js/src/access-rights/controllers/access-rights-modal-ctrl.js]: Should add a check that exits `vm.submit()` if the boolean is true
```js
vm.submit = function () {
  if (vm.submitting) {
    return;
  }

  // ...
}

File: contactaccessrights/js/src/access-rights/controllers/access-rights-modal-ctrl.js (hunk 1 above)
Comment (sleekvick) [inline contactaccessrights/js/src/access-rights/controllers/access-rights-modal-ctrl.js]: Sure...

File: contactaccessrights/js/src/access-rights/controllers/access-rights-modal-ctrl.js [hunk 4]
```diff
@@ -29,6 +32,14 @@ define([
       vm.dataLoaded = false;
 
       /**
+       * Change save button state
+       */
+      vm.selection = function () {
+        if (vm.selectedData.locations.length > 0 && vm.selectedData.regions.length > 0)
```
Comment (deb1990) [inline contactaccessrights/js/src/access-rights/controllers/access-rights-modal-ctrl.js]: use `{` with if

File: contactaccessrights/js/src/access-rights/controllers/access-rights-modal-ctrl.js [hunk 5]
```diff
@@ -9,6 +9,9 @@ define([
     function ($q, $modalInstance, Region, Location, Right) {
       var vm = this;
 
+      vm.submitting = false;
```
Comment (deb1990) [inline contactaccessrights/js/src/access-rights/controllers/access-rights-modal-ctrl.js]: please give meaningful names to the variables, right now both looks similar

File: contactaccessrights/js/test/controllers/access-rights-modal-ctrl_test.js [hunk 6]
```diff
@@ -223,6 +223,31 @@ define([
...
+            ctrl.selectedData.regions.push({
+              name: "Home",
+              label: "Home"
+            });
+
+            ctrl.selection();
+
+            expect(ctrl.canSubmit).toBe(true);
+          });
+
```
Comment (deb1990) [inline contactaccessrights/js/test/controllers/access-rights-modal-ctrl_test.js]: Remove extra empty lines

File: contactaccessrights/js/test/controllers/access-rights-modal-ctrl_test.js [hunk 7]
```diff
@@ -223,6 +223,31 @@ define([
           });
         });
 
+        describe("when location changes", function () {
+          it("cannot submit without change", function () {
+            expect(ctrl.canSubmit).toBe(false);
+          });
+
```
Comment (deb1990) [inline contactaccessrights/js/test/controllers/access-rights-modal-ctrl_test.js]: Remove extra empty lines

File: contactaccessrights/js/test/controllers/access-rights-modal-ctrl_test.js [hunk 8]
```diff
@@ -223,6 +223,31 @@ define([
...
+          it("can submit when location and region changes", function () {
```
Comment (deb1990) [inline contactaccessrights/js/test/controllers/access-rights-modal-ctrl_test.js]: Move the setup of the test case to beforeeach, the it block should contain only the expect statement

```
describe('when location and region changes', function () {
  
  beforeEach(function () {
    ctrl.selectedData.locations.push({
      name: "Home",
      label: "Home"
    });

    ctrl.selectedData.regions.push({
      name: "Home",
      label: "Home"
    });

    ctrl.selection();
  });
  
  it("can submit", function () {
    expect(ctrl.canSubmit).toBe(true);
  });
});
```

File: contactaccessrights/js/test/controllers/access-rights-modal-ctrl_test.js [hunk 9]
```diff
@@ -222,6 +222,45 @@ define([
...
         });
+
+        describe("when location changes", function () {
+          it("cannot submit without change", function () {
+            expect(ctrl.canSubmit).toBe(false);
+          });
+
+          it("can submit when location and region changes", function () {
+
+            ctrl.selectedData.locations.push({
```
Comment (deb1990) [inline contactaccessrights/js/test/controllers/access-rights-modal-ctrl_test.js]: You are not testing when either of `vm.selectedData.locations.length` or `vm.selectedData.regions.length` is false

File: contactaccessrights/js/test/controllers/access-rights-modal-ctrl_test.js [hunk 10]
```diff
@@ -222,6 +222,45 @@ define([
...
+          beforeEach(function () {
+            ctrl.submit();
+          });
+
+          it('submitting should be true', function () {
+            expect(ctrl.submitting).toBe(true);
+          });
+
+          xit('submitting should be false when promise has finished', function () {
+            // TODO
```
Comment (deb1990) [inline contactaccessrights/js/test/controllers/access-rights-modal-ctrl_test.js]: Please complete this test case

File: contactaccessrights/js/src/access-rights/controllers/access-rights-modal-ctrl.js [hunk 11]
```diff
@@ -39,12 +50,20 @@ define([
        * Saves data and closes the modal
        */
       vm.submit = function () {
+        if (vm.submitting) {
```
Comment (deb1990) [inline contactaccessrights/js/src/access-rights/controllers/access-rights-modal-ctrl.js]: Already we have ng-disabled specified, so we can skip this if logic

File: contactaccessrights/js/test/controllers/access-rights-modal-ctrl_test.js (hunk 10 above)
Comment (sleekvick) [inline contactaccessrights/js/test/controllers/access-rights-modal-ctrl_test.js]: No need for this test case, If it resolves or rejects the variable will be reinitialized.

**Pull request merged**